src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -279,58 +265,41 @@ void PhaseIdealLoop::clone_skeleton_predicates_to_unswitched_loop(Deoptimization
-          // Update the control input and reset the mapping for 'slow_node' back to NULL.
-          _igvn.replace_input_of(slow_node, 0, cloned_proj);
-          old_new.map(slow_node->_idx, NULL);
-          --i;
-        }
-        assert(slow_node->_idx <= idx_before_clone || old_new[slow_node->_idx] == NULL, "mapping of cloned nodes must be null");
-      }
-    } else {
-      // Fast loop
-      for (DUIterator i = predicate->outs(); predicate->has_out(i); i++) {
-        Node* fast_node = predicate->out(i);
+    _igvn.replace_input_of(fast_proj->in(0), 1, iff->in(1));
+    _igvn.replace_input_of(slow_proj->in(0), 1, iff->in(1));
+
+    for (DUIterator i = predicate->outs(); predicate->has_out(i); i++) {
+      Node* fast_node = predicate->out(i);
+      if (loop->is_member(get_loop(ctrl_or_self(fast_node)))) {
+        assert(fast_node->in(0) == predicate, "only control edge");
         Node* slow_node = old_new[fast_node->_idx];
-        if (!fast_node->is_CFG() && slow_node != NULL && slow_node->_idx > idx_before_clone) {
-          // 'fast_node' is a data node and part of the fast loop. Add the clone of the fast loop node
-          // to the 'old_new' mapping in order to verify later when cloning the predicates for the slow loop
-          // that 'slow_node' is a clone of 'fast_node'. Update the control input for 'fast_node'.
-          _igvn.replace_input_of(fast_node, 0, cloned_proj);
-          assert(old_new[slow_node->_idx] == NULL, "mapping must be null for cloned nodes");
-          old_new.map(slow_node->_idx, fast_node);
-          --i;
-        }
+        assert(slow_node->in(0) == predicate, "only control edge");
+        _igvn.replace_input_of(fast_node, 0, fast_proj);
+        to_process.push(slow_node);
+        --i;
       }
     }
+    // Have to delay updates to the slow loop so uses of predicate are
+    // not modified while we iterate on them.
+    while (to_process.size() > 0) {
+      Node* slow_node = to_process.pop();
+      _igvn.replace_input_of(slow_node, 0, slow_proj);
+    }
   }
 }
 
 //--------------------------clone_loop_predicates-----------------------
 // Clone loop predicates to cloned loops when unswitching a loop.
-Node* PhaseIdealLoop::clone_predicates_to_unswitched_loop(Node* old_entry, Node* new_entry, bool clone_limit_check,
-                                            bool is_slow_loop, uint idx_before_clone, Node_List &old_new) {
-#ifdef ASSERT
-  assert(LoopUnswitching, "sanity - only called when unswitching a loop");
-  if (new_entry == NULL || !(new_entry->is_Proj() || new_entry->is_Region() || new_entry->is_SafePoint())) {
-    if (new_entry != NULL)
-      new_entry->dump();
-    assert(false, "not IfTrue, IfFalse, Region or SafePoint");
-  }
-#endif
+void PhaseIdealLoop::clone_predicates_to_unswitched_loop(IdealLoopTree* loop, const Node_List& old_new, ProjNode*& iffast, ProjNode*& ifslow) {
+  LoopNode* head  = loop->_head->as_Loop();
```

Comment:
Extra whitespace after `head`

Reply:
I'll remove the white space before I integrate the change.